File: src/main/kotlin/inaction/chap4/innernested/java/Button.java
package inaction.chap4.innernested.java;

import inaction.chap4.innernested.State;
import inaction.chap4.innernested.View;
import org.jetbrains.annotations.NotNull;

public class Button implements View {
    @NotNull
    @Override
    public State getCurrentState() {
        return new ButtonState();
    }

    @Override
    public void restoreState(@NotNull State state) {/* ... */}

    public static class ButtonState implements State {
        // 여기에 Button 에 대한 구체적인 정보를 추가
    }
}


class Outer {
    int field1 = 0;
    class Inner {
        int field1 = 1;
        Outer getOuterReference() {return Outer.this;}

        int getOuterField1() {return Outer.this.field1;}
    }
}

File: src/main/kotlin/inaction/chap4/innernested/kotlin/Button.kt
package inaction.chap4.innernested.kotlin

import inaction.chap4.innernested.State
import inaction.chap4.innernested.View

class Button : View {
    override fun getCurrentState(): State = ButtonState()

    inner class ButtonState : State {
        // 여기에 Button 에 대한 구체적인 정보를 추가
    }
}


class Outer() {
    val field1: Int = 0
    inner class Inner() {
        val field1: Int = 1
        fun getOuterReference(): Outer = this@Outer
        fun getOuterField1(): Int = this@Outer.field1
    }
}